### PR TITLE
metricbeat: add integration TestIndexTotalFieldsLimitNotReached

### PR DESCRIPTION
**Pull request by AndersonQ**
## Proposed commit message

```
metricbeat: add integration TestIndexTotalFieldsLimitNotReached

The new integration test TestIndexTotalFieldsLimitNotReached ensures events with at least 500 new dynamically mapped fields can be ingested.
```

This new test will start failing when the metricbeat mapping has increased to the point it isn't possible to add a new event with 500 dynamically mapped fields. It should act as a safe guard to prevent us from releasing a version where the static mapping is too close to the total fields limit.

## Checklist

- [x] My code follows the style guidelines of this project
- ~~[ ] I have commented my code, particularly in hard-to-understand areas~~
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- ~~[ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~~

## Disruptive User Impact

- N/A

## How to test this PR locally

 - build `x-pack/metricbeat`
   - `cd x-pack/metricbeat`
   - `go build .`
 - adjust `metricbeat.yml` to point to your stack and make sure to include `setup.kibana.host`
 - run `./metricbeat setup`
 - generate an event with 500 dynamic fields. Just use the same for loop the test uses
 - ingest the event:
```
POST metricbeat-9.0.0/_doc
{ 
   ...
}
```
 - it should succeed
 - change the [`index.mapping.total_fields.limit`](https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-settings-limit.html) for the metricbeat index template:
   - go to `your-kibna/app/management/data/index_management/templates/metricbeat-9.0.0`
   - `metricbeat-9.0.0` > Manage > Edit > 3 Index settings
   - change `index.mapping.total_fields.limit` to 10000:
```json
"mapping": {
      "total_fields": {
        "limit": "10000"
      }
```
 - rollover the index
```
POST metricbeat-9.0.0/_rollover
```
 - try to ingest the same event again, you should get an error like:
```
{
  "error": {
    "root_cause": [
      {
        "type": "document_parsing_exception",
        "reason": "[1:378] failed to parse: Limit of total fields [10000] has been exceeded while adding new fields [11]"
      }
    ],
    "type": "document_parsing_exception",
    "reason": "[1:378] failed to parse: Limit of total fields [10000] has been exceeded while adding new fields [11]",
    "caused_by": {
      "type": "illegal_argument_exception",
      "reason": "Limit of total fields [10000] has been exceeded while adding new fields [11]"
    },
    "failure_store": "not_enabled"
  },
  "status": 400
}
```

## Related issues

- Relates https://github.com/elastic/beats/pull/41614

### CODE REVIEW

File: x-pack/metricbeat/tests/integration/setup_integration_test.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+		"--index-management",
+		"-E", "setup.kibana.protocol=http",
+		"-E", "setup.kibana.host="+kURL.Hostname(),
+		"-E", "setup.kibana.port="+kURL.Port(),
+		"-E", "output.elasticsearch.protocol=http",
+		"-E", "output.elasticsearch.hosts=['"+esURL.String()+"']",
+		"-E", "output.file.enabled=false")
+	procState, err := metricbeat.Process.Wait()
+	require.NoError(t, err, "metricbeat setup failed")
+	require.Equal(t, 0, procState.ExitCode(), "metircbeat setup failed: incorrect exit code")
```
Comment (mauri870) [inline x-pack/metricbeat/tests/integration/setup_integration_test.go:71]:
```suggestion
	require.Equal(t, 0, procState.ExitCode(), "metricbeat setup failed: incorrect exit code")
```

File: x-pack/metricbeat/tests/integration/setup_integration_test.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+	r, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, bytes.NewBuffer(event))
+	require.NoError(t, err, "could not create request to send event to ES")
+	r.Header.Set("Content-Type", "application/json")
+
+	resp, err := http.DefaultClient.Do(r)
+	require.NoError(t, err, "could not send request to send event to ES")
+	defer resp.Body.Close()
+
+	failuremsg := fmt.Sprintf("filed to ingest events with %d new fields. If this test fails it likely means the current `index.mapping.total_fields.limit` for metricbeat index (%s) is close to be reached. Check the logs to see why the envent was not ingested", totalFields, index)
```
Comment (mauri870) [inline x-pack/metricbeat/tests/integration/setup_integration_test.go:98]:
```suggestion
	failuremsg := fmt.Sprintf("failed to ingest events with %d new fields. If this test fails it likely means the current `index.mapping.total_fields.limit` for metricbeat index (%s) is close to be reached. Check the logs to see why the event was not ingested", totalFields, index)
```

File: x-pack/metricbeat/tests/integration/setup_integration_test.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	if !assert.Equal(t, http.StatusCreated, resp.StatusCode, failuremsg) {
```
Comment (mauri870) [inline x-pack/metricbeat/tests/integration/setup_integration_test.go:99]: Could use assert.NotEqual here

**Review by mauri870 — APPROVED** [review]
LGTM Left some comments, but overall the test seems correct to me.

The linter error seems to be the same one that Lee faced this week, caused by an old version of golangci-lint.

File: x-pack/metricbeat/tests/integration/setup_integration_test.go (hunk 3 above)
Comment (AndersonQ) [reply x-pack/metricbeat/tests/integration/setup_integration_test.go:99]: not really. I want to assert it's equal, if the assertion fails, then the extra code runs to help to explain what went wrong

File: x-pack/metricbeat/tests/integration/setup_integration_test.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	cleanUpES := func() {
+		_, _, err := integration.HttpDo(t, http.MethodDelete, dataStreamURL)
+		require.NoErrorf(t, err, "cleanup failed: could not remove datastream %s", index)
+		_, _, err = integration.HttpDo(t, http.MethodDelete, templateURL)
+		require.NoErrorf(t, err, "cleanup failed: could not remove index template %s", index)
+		_, _, err = integration.HttpDo(t, http.MethodDelete, policyURL)
+		require.NoErrorf(t, err, "cleanup failed: could not ilm policy %s", index)
+	}
+	// ensure datastream/index template/ ilm policy is set before running the test
+	cleanUpES()
```
Comment (belimawr) [inline x-pack/metricbeat/tests/integration/setup_integration_test.go:58]: What do you mean?

You're removing them all, however you wrote about ensuring they're set... I'm confused on what you're trying to accomplish here.

File: x-pack/metricbeat/tests/integration/setup_integration_test.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	t.Cleanup(cleanUpES)
+
+	metricbeat.Start("setup",
+		"--index-management",
+		"-E", "setup.kibana.protocol=http",
+		"-E", "setup.kibana.host="+kURL.Hostname(),
+		"-E", "setup.kibana.port="+kURL.Port(),
+		"-E", "output.elasticsearch.protocol=http",
+		"-E", "output.elasticsearch.hosts=['"+esURL.String()+"']",
+		"-E", "output.file.enabled=false")
```
Comment (belimawr) [inline x-pack/metricbeat/tests/integration/setup_integration_test.go:68]: Why do you need to set the file output to false?
```suggestion
```

File: x-pack/metricbeat/tests/integration/setup_integration_test.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+	procState, err := metricbeat.Process.Wait()
+	require.NoError(t, err, "metricbeat setup failed")
+	require.Equal(t, 0, procState.ExitCode(), "metricbeat setup failed: incorrect exit code")
```
Comment (belimawr) [inline x-pack/metricbeat/tests/integration/setup_integration_test.go:71]: [Suggestion]
Add the exit code to the error message.
```suggestion
	require.Equalf(t, 0, procState.ExitCode(), "metricbeat setup failed: incorrect exit code: %d", procState.ExitCode())
```

File: x-pack/metricbeat/tests/integration/setup_integration_test.go [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+	// generate an event with dynamically mapped fields
+	fields := map[string]string{}
+	totalFields := 500
+	for i := range totalFields {
+		fields[fmt.Sprintf("a-label-%d", i)] = fmt.Sprintf("some-value-%d", i)
+	}
+	event, err := json.Marshal(map[string]any{
+		"@timestamp":        time.Now().Format(time.RFC3339),
+		"kubernetes.labels": fields,
```
Comment (belimawr) [inline x-pack/metricbeat/tests/integration/setup_integration_test.go:81]: Why `kubernete.labels`? Wouldn't the same happen on any field? It it has to be `kubernetes.labels` it would be nice to add a comment explaining why.

**Comment by cmacknz** [comment]
Just to confirm, if you revert https://github.com/elastic/beats/pull/41640 this test fails immediately?

File: x-pack/metricbeat/tests/integration/setup_integration_test.go [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+	r, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, bytes.NewBuffer(event))
+	require.NoError(t, err, "could not create request to send event to ES")
+	r.Header.Set("Content-Type", "application/json")
+
+	resp, err := http.DefaultClient.Do(r)
+	require.NoError(t, err, "could not send request to send event to ES")
+	defer resp.Body.Close()
+
+	failuremsg := fmt.Sprintf("failed to ingest events with %d new fields. If this test fails it likely means the current `index.mapping.total_fields.limit` for metricbeat index (%s) is close to be reached. Check the logs to see why the event was not ingested", totalFields, index)
```
Comment (cmacknz) [inline x-pack/metricbeat/tests/integration/setup_integration_test.go:99]:
```suggestion
	failuremsg := fmt.Sprintf("failed to ingest events with %d new fields. If this test fails it likely means the current `index.mapping.total_fields.limit` for metricbeat index (%s) is close to being reached. Check the logs to see why the event was not ingested", totalFields, index)
```

File: x-pack/metricbeat/tests/integration/setup_integration_test.go [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/elastic/beats/v7/libbeat/tests/integration"
+	"github.com/elastic/beats/v7/libbeat/version"
+)
+
+func TestIndexTotalFieldsLimitNotReached(t *testing.T) {
```
Comment (cmacknz) [inline x-pack/metricbeat/tests/integration/setup_integration_test.go:27]: I would add some explicit documentation for why this test exists and that it is a regression test for https://github.com/elastic/beats/pull/41614 specifically.

This test is likely to fail for seemingly unrelated changes so let's go overboard with documenting the problem it is detecting.

**Comment by cmacknz** [comment]
> Just to confirm, if you revert https://github.com/elastic/beats/pull/41640 this test fails immediately?

I see this was answered in the description reading it again 👍
